`TestMonitorStaleLocks ` flakiness

## Failure

#### Link(s) to logs

- https://github.com/gravitational/teleport/actions/runs/5073276431/jobs/9112079298

#### Relevant snippet

```
===================================================
7443 tests passed. 2 failed, 122 skipped
===================================================
FAIL: github.com/gravitational/teleport/lib/srv
FAIL: github.com/gravitational/teleport/lib/srv.TestMonitorStaleLocks
===================================================
OUTPUT github.com/gravitational/teleport/lib/srv.TestMonitorStaleLocks
===================================================
=== RUN   TestMonitorStaleLocks
=== PAUSE TestMonitorStaleLocks
=== CONT  TestMonitorStaleLocks
{"caller":"lite/lite.go:246","component":"sqlite","level":"debug","message":"Connected to: file:%2Ftmp%2FTestEmitExecAuditEvent2810951033%2F001%2Fsqlite.db?_busy_timeout=10000\u0026_sync=FULL\u0026_txlock=immediate, poll stream period: 1s","timestamp":"2023-05-24T20:51:17Z"}
    monitor_test.go:223: Timeout waiting for LockWatcher reset.
--- FAIL: TestMonitorStaleLocks (15.42s)
{"caller":"services/watcher.go:243","component":null,"level":"debug","message":"Closed, returning from watch loop.","resource-kind":"lock","timestamp":"2023-05-24T20:51:32Z"}
```

--- a/lib/services/watcher.go
+++ b/lib/services/watcher.go
@@ -242,6 +242,11 @@ func (p *resourceWatcher) runWatchLoop() {
 		case <-p.ctx.Done():
 			p.Log.Debug("Closed, returning from watch loop.")
 			return
+		case <-p.StaleC:
+			// Used for testing that the watch routine is waiting for the
+			// next restart attempt. We don't want to wait for the full
+			// retry period in tests so we trigger the restart immediately.
+			p.Log.Debug("Stale view, continue watch loop.")
 		}
 		if err != nil {
 			p.Log.Warningf("Restart watch on error: %v.", err)

--- a/lib/srv/monitor_test.go
+++ b/lib/srv/monitor_test.go
@@ -205,9 +205,6 @@ func TestMonitorStaleLocks(t *testing.T) {
 	case <-time.After(15 * time.Second):
 		t.Fatal("Timeout waiting for LockWatcher loop check.")
 	}
-	// StaleC is listened by multiple goroutines, so we need to close to ensure
-	// that all of them are unblocked.
-	close(asrv.LockWatcher.StaleC)
 
 	// ensure ResetC is drained
 	select {
@@ -222,6 +219,12 @@ func TestMonitorStaleLocks(t *testing.T) {
 	case <-time.After(15 * time.Second):
 		t.Fatal("Timeout waiting for LockWatcher reset.")
 	}
+	// StaleC is listened by multiple goroutines, so we need to close to ensure
+	// that all of them are unblocked and the stale state is detected.
+	close(asrv.LockWatcher.StaleC)
+	require.Eventually(t, func() bool {
+		return asrv.LockWatcher.IsStale()
+	}, 15*time.Second, 100*time.Millisecond, "Timeout waiting for LockWatcher to be stale.")
 	select {
 	case <-conn.closedC:
 	case <-time.After(15 * time.Second):

